packages/arb-token-bridge-ui/src/tokenLists.tsx

```diff
@@ -123,29 +115,35 @@ export async function fetchTokenListFromURL(
 
     if (!validateTokenList(data)) {
       console.warn('Token List Invalid', data)
-      return { isValid: false, data }
+      return { isValid: false, data, tokenListURL }
     }
 
-    return { isValid: true, data }
+    return { isValid: true, data, tokenListURL }
   } catch (error) {
     console.warn('Token List URL Invalid', tokenListURL)
-    return { isValid: false, data: undefined }
+    return { isValid: false, data: undefined, tokenListURL }
   }
 }
 
-let storage: string | null = null
-export function fetchTokenLists(): Promise<void> {
+export function fetchTokenLists(
+  forL2ChainId: number
+): Promise<TokenListWithId[]> {
   return new Promise(resolve => {
     Promise.all(
-      BRIDGE_TOKEN_LISTS.map(bridgeTokenList =>
-        fetchTokenListFromURL(bridgeTokenList.url)
-      )
+      BRIDGE_TOKEN_LISTS.filter(
+        bridgeTokenList => bridgeTokenList.originChainID === forL2ChainId
+      ).map(bridgeTokenList => fetchTokenListFromURL(bridgeTokenList.url))
     ).then(responses => {
       const tokenListsWithBridgeTokenListId = responses
         .filter(({ isValid }) => isValid)
         // Attach the bridge token list id so we can easily retrieve a list later
-        .map(({ data }, index) => {
-          const token = BRIDGE_TOKEN_LISTS[index]
+        .map(({ data, tokenListURL }) => {
+          // TODO: this can be better handled if we receive 1-to-1 id from token-lists in response as well
```

Comment:
If we first `map`, THEN `filter`, we can access it by index. It should not be a big issue as we don't do anything expensive in the `map`. You'll probably need to extract the L133-135 to a variable, so we can access it by index